README.md
# hakatime

[![Latest version](https://img.shields.io/github/v/release/mujx/hakatime)](https://github.com/mujx/hakatime/releases)
[![BuiltWithNix](https://img.shields.io/badge/Built_With-Nix-5277C3.svg?logo=nixos&labelColor=73C3D5)](https://nixos.org/)

Hakatime is a server implementation of [Wakatime](https://wakatime.com/). You can use it as a self-hosted alternative.

It comes together with a dashboard which provides a graphical representation of the collected data.

## Features

- Import Wakatime activity using an API token and a range of dates.
- See time spent on Github commits.
- Forward incoming heartbeats to another Wakatime compatible server (e.g `wakatime.com`)
- Group projects together with tags (e.g `#work`, `#personal`) and view their aggregated statistics.
- User registration & login through the UI.
- Leaderboards for all the users of the instance.
- Badge generation for a project that displays that total amount of hours spent for a configurable
  time period. [![my-app](https://hakatime.mtx-dev.xyz/badge/svg/4e8273d1-84bc-46f9-890b-14fdfb83cdcc?days=14)](https://hakatime.mtx-dev.xyz/badge/svg/4e8273d1-84bc-46f9-890b-14fdfb83cdcc?days=7)
- Global and per project charts
  - Breakdown by project or language.
  - Breakdown by day of week and hour of the day.
  - Timeline of activity for a configurable time-frame.
  - Total time spent per file.
- API token management & generation.

## Demo

There is an instance [here](https://hakatime.mtx-dev.xyz) that you can use to send data or browse
around.

Use `demo` as username and `demodemo` as password or create your own user to login. You can create
an API token and try sending your data.

## Client setup

After you've generated a new API token from Hakatime's UI, update your `~/.wakatime.cfg` file like below:

```ini
[settings]
api_key=<generated_token>

# e.g api_url=https://hakatime.mtx-dev.xyz/api/v1/users/current/heartbeats.bulk for the demo instance.
api_url=<hakatime_instance_url>/api/v1/users/current/heartbeats.bulk

# e.g If you're using the docker-compose-deploy.yml to test things locally, use the following:
# api_url=http://localhost:8080/api/v1/users/current/heartbeats.bulk
```

The file's location might be different depending on your system and editor. Please consult the [client docs](https://wakatime.com/plugins)
for further information.

## Deployment

You can use the following docker-compose setup for testing locally or an actual
deployment. Change `HAKA_BADGE_URL` to match the actual external endpoint of your
instance.

Deploying on ARM is also possible using the `mujx/hakatime:latest-arm` or `mujx/hakatime:v1.7.3-arm` image or the dedicated
Dockerfile ([`Dockerfile.arm`](/Dockerfile.arm)) to build the image.

```yaml
version: "3"
services:
  server:
    container_name: hakatime
    image: mujx/hakatime:v1.7.3
    environment:
      # DB settings.
      HAKA_DB_HOST: haka_db
      HAKA_DB_PORT: 5432
      HAKA_DB_NAME: test
      HAKA_DB_PASS: test
      HAKA_DB_USER: test
      # Server settings.
      # Fill out this field if the api is behind another path (e.g behind a reverse proxy).
      # This will adjust the Set-Cookie path for all the /auth related API calls.
      HAKA_API_PREFIX: ""
      # Update this with the external endpoint that you use to access hakatime.
      HAKA_BADGE_URL: "http://localhost:8080"
      HAKA_PORT: 8080
      HAKA_SHIELDS_IO_URL: "https://img.shields.io"
      HAKA_ENABLE_REGISTRATION: "true" # Toggle after you've created your account.
      # Number of hours after which inactive browser sessions will expire (login required).
      HAKA_SESSION_EXPIRY: "24"
      HAKA_LOG_LEVEL: "info" # Control the verbosity of the logger.
      HAKA_ENV: "dev" # Use a json logger for production, otherwise key=value pairs.
      HAKA_HTTP_LOG: "true" # If you want to log http requests.
      GITHUB_TOKEN: "<token>" # If you want to retrieve time spent per commit. No extra scope is required.
      # Add the following variables if you want to forward any received heartbeats to another
      # Wakatime compatible server.
      HAKA_REMOTE_WRITE_URL: "https://wakatime.com/api/v1/users/current/heartbeats.bulk"
      HAKA_REMOTE_WRITE_TOKEN: "<token>"
    ports:
      - "127.0.0.1:8080:8080"
  haka_db:
    container_name: haka_db
    image: postgres:12-alpine
    environment:
      POSTGRES_DB: test
      POSTGRES_PASSWORD: test
      POSTGRES_USER: test
    volumes:
      - deploy_db_data:/var/lib/postgresql/data

volumes:
  deploy_db_data: {}
```

To start all the services run:

```bash
$ docker-compose -f ./docker-compose-deploy.yml up
```

and navigate to [http://localhost:8080](http://localhost:8080) to access the dashboard.

## Building

### Server

Requirements:

- [GHC](https://www.haskell.org/ghc/) (tested with 8.8 & 8.10)
- [libpq](https://www.postgresql.org/docs/11/libpq.html) (PostgreSQL bindings)
- [cabal-install](https://www.haskell.org/cabal/) (Build system for Haskell)

```bash
cabal build
cabal run exe:hakatime -- run
```

### Dashboard

The output files will be located at `dashboard/dist`.

Requirements:

- Node.js & npm

```bash
cd dashboard

npm install
npm run prod
```

## CLI options

```
hakatime :: v1.7.3

Usage: hakatime COMMAND
  Wakatime server implementation

Available options:
  -h,--help                Show this help text

Available commands:
  create-token             Create a new auth token
  create-user              Create a new user account
  run-migrations           Apply pending database migrations
  run                      Start the Server
```

## Screens

### Overview

![Overview Page](img/overview.png "Overview Page")

### Projects

![Projects Page](img/projects.png "Projects Page")

![Projects Page - Commits](img/commits.png "Projects Page - Commits")

## Contributing

Any kind of contribution is greatly appreciated. This could be:

- Bug fixes
- Suggesting/Implementing new features
- UI/UX improvements/suggestions
- Code refactoring


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.7.3] - 2023-04-13

### Bug fixes

- Fix off by one error when grouping activity by hour (@ablearthy #74)
- Add psql client on the ARM image (#71)

## [1.7.2] - 2022-09-17

### Bug fixes

- Convert projects with empty name to 'Unknown project' (#62)

## [1.7.1] - 2022-04-10

### Bug fixes

- Add `url` heartbeat type (thanks @georgy-komarov, #59)

## [1.7.0] - 2021-10-25

### Features

- Add endpoint to show the time spent on the current day on the VS Code statusbar. (#50)

## [1.6.3] - 2021-10-16

This release should fix the issues encountered using the new wakatime-cli.

### Bug fixes

- Use `Int64` to represent the `cursorpos` field.
- Respond with the correct response format when single heartbeats are sent. (#49)

## [1.6.2] - 2021-10-10

### Bug fixes

- Parse the `cursorpos` heartbeat field as either `Text` or `Int` (#46)
- Reply with 202 status code when new heartbeats are inserted (#48)

## [1.6.1] - 2021-08-31

### Bug fixes

- Switched from `api.wakatime.com` to `wakatime.com` during data import. The former endpoint was
  returning HTML instead of JSON and wasn't working over HTTPS. (#43)

## [1.6.0] - 2021-08-31

### Features

- Add ability to view aggregated project data with tags (instead of aliases) (#32)
  To use, set a common tag on the projects you want to view as one and select the tag on the
  Projects page.

## [1.5.0] - 2021-08-08

### Features

- Add ability to set names on API tokens (#38)

## [1.4.1] - 2021-06-14

### Bug fixes

- Allow remote write http endpoints and custom port (#36)
- Install the `ca-certificates` package in the arm Dockerfile (#36)

## [1.4.0] - 2021-06-11

### Features

- Add ability to forward incoming heartbeats to another Wakatime compatible server.

### Improvements

- Docker images for ARM64 are built automatically by CircleCI.

## [1.3.2] - 2021-04-11

### Bug fixes

- Update project list according to the selected date range.

## [1.3.1] - 2021-04-04

### Improvements

- Don't display coding time for merge commits.
- Show hours & minutes on all charts.
- Update some dashboard dependencies.

## [1.3.0] - 2021-03-19

### Features

- Added API endpoint & UI modal to see time spent per GitHub commit.

### Improvements

- Converted all tooltips to use the Bootstrap theme.
- Added explanation about the cut-off limit (renamed to timeout).

## [1.2.0] - 2021-03-14

### Features

- Added leaderboards for the users of the instance.

### Improvements

- Small visual improvements on the login & register pages.
- Removed very low percentages from pie charts to reduce noise.

## [1.1.0] - 2021-03-01

### Features

- API endpoint & UI controls to attach tags to projects. (e.g `work`, `web`, `ui`)
- Ability to filter projects by tag and view their aggregated statistics.

### Improvements

- Reduced the amount of api calls made by the UI.
- Error popups on the UI when queries fail.
- The docker image will apply the database migrations on boot.

### Bug fixes

- Removed a small UI freeze by not redrawing immediately upon dropdown selection.

## [1.0.1] - 2021-02-23

### Improvements

- The following file extensions are now recognized (the wakatime plugin sends no values for these):
  - `.gotmpl` -> `Go template`
  - `.tfvars` -> `Terraform`
  - `.dhall` -> `Dhall`
  - `.zig` -> `Zig`
  - `.org` -> `Org`
  - `.purs` -> `PureScript`
  - `.cabal` -> `Cabal config`
  - `.jinja|.jinja2` -> `Jinja`
- Updated the docker image to use GHC 8.10.
- Reduced the size of the docker image down to ~85 MB.
- Removed some unnecessary direct dependencies.

### Bug fixes

- Fixed the Dockerfile for ARM (#21). (TravisCI experiences random failures and it's currently not possible to
  provide regular docker builds for ARM)

## [1.0.0] - 2021-02-16

Initial versioned release.

### Features

- Import Wakatime activity using an API token and a range of dates.
- User registration & login through the UI.
- Badge generation for a project that displays that total amount of hours spent for a configurable
  time period.
- Global and per project charts
  - Breakdown by project or language.
  - Breakdown by day of week and hour of the day.
  - Timeline of activity for a configurable time-frame.
  - Total time spent per file.
- API token management & generation.


app/Main.hs
module Main
  ( main,
  )
where

import Control.Concurrent (forkIO)
import Control.Exception.Safe (catchAny, try)
import qualified GHC.IO.Encoding
import qualified Haka.Api as Api
import Haka.App
  ( AppCtx (..),
    AppM,
    LogState (..),
    RegistrationStatus (..),
    RemoteWriteConfig (..),
    ServerSettings (..),
    runAppT,
  )
import qualified Haka.Cli as Cli
import qualified Haka.Handlers.Import as Import
import qualified Haka.Logger as Log
import qualified Haka.Middleware as Middleware
import qualified Hasql.Connection as HasqlConn
import qualified Hasql.Pool as HasqlPool
import Hasql.Queue.Migrate (migrate)
import Katip
import Network.Wai
import Network.Wai.Handler.Warp
import Network.Wai.Logger (withStdoutLogger)
import Network.Wai.Middleware.Cors
import qualified Options.Applicative as Opt
import Servant
import System.Environment.MrEnv (envAsBool, envAsInt, envAsString)
import System.Posix.Env.ByteString (getEnvDefault)

-- | Convert our 'App' type to a 'Servant.Handler', for a given 'AppCtx'.
nt :: AppCtx -> AppM a -> Handler a
nt ctx = Handler . ExceptT . try . runAppT ctx

app :: ServerSettings -> AppCtx -> Application
app settings conf =
  Middleware.jsonResponse $
    cors (const $ Just policy) $
      serve Api.api $
        hoistServer Api.api (nt conf) (Api.server settings)
  where
    policy = simpleCorsResourcePolicy

initApp :: ServerSettings -> (AppCtx -> Application) -> IO ()
initApp settings unApp = do
  dbSettings <- Cli.getDbSettings
  dbPool <- HasqlPool.acquire 10 Nothing dbSettings

  -- Set up the db schema for the postgres queue.
  res <- HasqlConn.acquire dbSettings
  case res of
    Left e -> do
      let errMsg = "Failed to connect to the database:\n"
      case e of
        Just errS -> die $ errMsg <> decodeUtf8 errS
        Nothing -> die errMsg
    Right conn -> migrate conn "json"

  logenv <-
    Log.setupLogEnv
      (hakaRunEnv settings)
      (fromMaybe InfoS (textToSeverity $ toText $ hakaLogLevel settings))
  let logState' =
        LogState
          { lsNamespace = Namespace {unNamespace = ["api"]},
            lsLogEnv = logenv,
            lsContext = mempty
          }
      appCtx =
        AppCtx
          { pool = dbPool,
            logState = logState',
            srvSettings = settings
          }

  -- Handle import requests on another thread.
  _ <-
    forkIO $
      forever $
        runAppT appCtx Import.handleImportRequest `catchAny` Import.logExceptions logenv

  if hakaHasHttpLogger settings
    then do
      withStdoutLogger $ \logger -> do
        let conf = setPort (hakaPort settings) $ setLogger logger defaultSettings
        runSettings conf (unApp appCtx)
    else do
      run (hakaPort settings) (unApp appCtx)

emptyStr :: String
emptyStr = ""

getServerSettings :: IO ServerSettings
getServerSettings = do
  p <- envAsInt "HAKA_PORT" 8080
  badgeUrl <- getEnvDefault "HAKA_BADGE_URL" "http://localhost:8080"
  dashboardPath <- envAsString "HAKA_DASHBOARD_PATH" "./dashboard/dist"
  apiPrefix <- envAsString "HAKA_API_PREFIX" emptyStr
  shieldsIOUrl <- envAsString "HAKA_SHIELDS_IO_URL" "https://img.shields.io"
  enableRegistration <- envAsBool "HAKA_ENABLE_REGISTRATION" True
  sessionExpiry <- envAsInt "HAKA_SESSION_EXPIRY" 24
  logLevel <- envAsString "HAKA_LOG_LEVEL" "info"
  rEnv <- envAsString "HAKA_ENV" "prod"
  enableHttpLog <- envAsBool "HAKA_HTTP_LOG" True
  remoteWriteUrl <- envAsString "HAKA_REMOTE_WRITE_URL" emptyStr
  remoteWriteToken <- envAsString "HAKA_REMOTE_WRITE_TOKEN" emptyStr
  when (sessionExpiry <= 0) (error "Session expiry should be a positive integer")
  return $
    ServerSettings
      { hakaPort = p,
        hakaApiPrefix = apiPrefix,
        hakaBadgeUrl = badgeUrl,
        hakaDashboardPath = dashboardPath,
        hakaShieldsIOUrl = shieldsIOUrl,
        hakaLogLevel = logLevel,
        hakaHasHttpLogger = enableHttpLog,
        hakaRunEnv = case rEnv of
          "prod" -> Log.Prod
          "production" -> Log.Prod
          _ -> Log.Dev,
        hakaRemoteWriteConfig =
          if remoteWriteUrl /= emptyStr && remoteWriteToken /= emptyStr
            then
              Just $
                RemoteWriteConfig
                  { heartbeatUrl = toText remoteWriteUrl,
                    token = toText remoteWriteToken
                  }
            else Nothing,
        hakaEnableRegistration =
          if enableRegistration
            then EnabledRegistration
            else DisabledRegistration,
        hakaSessionExpiry = fromIntegral sessionExpiry
      }

main :: IO ()
main = do
  GHC.IO.Encoding.setLocaleEncoding GHC.IO.Encoding.utf8

  parsedOpts <- Opt.execParser Cli.opts
  s <- getServerSettings
  Cli.handleCommand (Cli.serverCmd parsedOpts) (initApp s (app s))
